### DISCUSSION TITLE
spec is missing one or more constants

### DISCUSSION THREAD

**Discussion by gcptcgnye61** [cilium/ebpf#1039] [Q&A]
Greetings, I have met a weird problem. When trying to RewriteConstant of the predefined volatile constant IPV6_INNATE_ADDR. Cilium/eBPF reported spec is missing one or more constants: IPV6_INNATE_ADDR.  But I have used the constant in the following code:
```
static volatile const __be32 IPV6_INNATE_ADDR[4] = {0,0,0,0};
...
// bpf_printk("daddr = %x, IPV6_INNATE_ADDR[0] = %x",
//            encap_ip6.daddr.s6_addr32[0], IPV6_INNATE_ADDR[0]);
struct in6_addr *prefix = (struct in6_addr *)IPV6_INNATE_ADDR;
unsigned long long out = ipv6_addr_cmp(prefix, &encap_ip6.daddr);
bpf_printk("out = %d", out);
if (out != 0) {
  bpf_printk("Match failed");
  return TC_ACT_OK;
}
```
If I uncomment the bpf_printk portion of the code, the warning goes away. Is this because I casted IPV6_INNATE_ADDR? I am not sure if this is a optimization problem or ebpf-go problem.

@lmb think this is a problem with how clang deals with array accesses through a volatile pointer, and suggested that I could try to make `struct in6_addr *prefix` volatile as well. So I made the following changes: 
```
volatile struct in6_addr *prefix =
    (volatile struct in6_addr *)IPV6_INNATE_ADDR;
unsigned long long out =
    __bpf_memcmp_volatile(prefix, &encap_ip6.daddr, sizeof(struct in6_addr));
```

Unfortunatly this did not make the error go away. But by using struct instead of array when defining the constant, I avoided this problem: 
```
static volatile const struct in6_addr IPV6_RSS_PREFIX = {
    {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
      0x0}}};
```

Does anyone have idea why a volatile array have such problem? Thanks.

**Comment by ti-mo** [comment]
Since this is likely down to compiler behaviour (ebpf-go just rewrites the constants it finds, there's not much more it can do), your best bet is to put up a reproducer (a git repo with full code and clear instructions on how to run) and interact on the kernel mailing list(s) to get some eyes on this: https://www.kernel.org/doc/html/latest/bpf/bpf_devel_QA.html#q-reporting-llvm-bpf-issues.
